merlin/systems/dag/runtimes/nvtabular/executor.py

```diff
@@ -0,0 +1,305 @@
+            )
+
+        output_data = None
+
+        for node in nodes:
+            transformed_data = self._execute_node(node, transformable)
+            output_data = self._combine_node_outputs(node, transformed_data, output_data)
+
+        if additional_columns:
+            output_data = concat_columns(
+                [output_data, transformable[_get_unique(additional_columns)]]
+            )
+
+        format_ = _data_format(output_data)
+        if format_ != output_format:
+            output_data = _convert_format(output_data, output_format)
+
+        return output_data
+
+    def _execute_node(self, workflow_node, input_tensors, capture_dtypes=False, strict=False):
+        upstream_outputs = self._run_upstream_transforms(workflow_node, input_tensors)
+        upstream_outputs = self._merge_addl_root_columns(
+            workflow_node, input_tensors, upstream_outputs
+        )
+        tensors = self._standardize_formats(workflow_node, upstream_outputs)
+
+        transform_input = _concat_tensors(tensors)
+        # TODO: In order to replace the line above with the line below, we first have to replace
+        #       dictionaries with TensorTables
+        # transform_input = self._merge_upstream_columns(tensors, merge_fn=_concat_tensors)
+        transform_output = self._run_node_transform(workflow_node, transform_input)
+
+        return transform_output
+
+    def _merge_addl_root_columns(self, workflow_node, input_tensors, upstream_outputs):
+        if workflow_node.selector:
+            selector_columns = workflow_node.selector.names
+            to_remove = []
+            for upstream_tensors in upstream_outputs or []:
+                for col in selector_columns:
+                    if col in upstream_tensors:
+                        to_remove.append(col)
+            for col in set(to_remove):
+                selector_columns.remove(col)
+
+            if selector_columns:
+                selected_tensors = {c: input_tensors[c] for c in selector_columns}
+                upstream_outputs.append(selected_tensors)
+
+        return upstream_outputs
+
+    def _standardize_formats(self, workflow_node, node_input_data):
+        # Get the supported formats
+        op = workflow_node.op
+        if op and hasattr(op, "inference_initialize"):
+            supported_formats = _maybe_mask_cpu_only(op.supports, self.device)
+        else:
+            supported_formats = Supports.CPU_DICT_ARRAY
+
+        # Convert the first thing into a supported format
+        tensors = _convert_format(node_input_data[0], supported_formats)
+        target_format = _data_format(tensors)
+
+        # Convert the whole list into the same format
+        formatted_tensors = []
+        for upstream_tensors in node_input_data:
+            upstream_tensors = _convert_format(upstream_tensors, target_format)
+            formatted_tensors.append(upstream_tensors)
+
+        return formatted_tensors
+
+
```

Comment:
Note that pretty much everything in this file from here down is just getting moved from other places. There's definitely more refactoring we want to do on the code below, but moving it here more clearly communicates that the usage of this code is isolated to this single file and can be freely refactored without causing issues elsewhere.